### ISSUE TITLE
Error when copying / pasting one column of data; more than one column is ok

### ISSUE THREAD

**Issue by pybokeh** [ZoomerAnalytics/xlwings#398]
Hello, I have an Excel file with just one column of data.  When I attempt to copy that column and then paste the contents to a different column by using this assignment:  
    
    Range('E1').value = Range('A1').vertical.value
    
I get this error:

    ----------------------------------------------------------------------------------------
    com_error                                 Traceback (most recent call last)
    <ipython-input-71-1ff4d126bb3c> in <module>()
    ----> 1 Range('E1').value = Range('A1').vertical.value

    D:\Miniconda3\envs\jupyter\lib\site-packages\xlwings\main.py in value(self, data)
        847     @value.setter
        848     def value(self, data):
    --> 849         conversion.write(data, self, self._options)
        850
        851     @property

    D:\Miniconda3\envs\jupyter\lib\site-packages\xlwings\conversion\__init__.py in write(value, rng, options)
         20     pipeline = accessors.get(convert, convert).router(value, rng, options).writer(options)
         21     ctx = ConversionContext(rng=rng, value=value)
    ---> 22     pipeline(ctx)
         23     return ctx.value

    D:\Miniconda3\envs\jupyter\lib\site-packages\xlwings\conversion\framework.py in __call__(self, *args, **kwargs)
         64     def __call__(self, *args, **kwargs):
         65         for stage in self:
    ---> 66             stage(*args, **kwargs)
         67
         68

    D:\Miniconda3\envs\jupyter\lib\site-packages\xlwings\conversion\standard.py in __call__(self, ctx)
         53                 ctx.range.col1,
         54                 row2,
    ---> 55                 col2
         56             ), ctx.value)
         57

    D:\Miniconda3\envs\jupyter\lib\site-packages\xlwings\_xlwindows.py in get_range_from_indices(xl_sheet, first_row, first_column, last_row, last_column)
        283 def get_range_from_indices(xl_sheet, first_row, first_column, last_row, last_column):
        284     return xl_sheet.Range(xl_sheet.Cells(first_row, first_column),
    --> 285                           xl_sheet.Cells(last_row, last_column))
        286
        287

    C:\Users\MA17151\AppData\Local\Temp\gen_py\3.4\00020813-0000-0000-C000-000000000046x0x1x6\Range.py in __call__(self, RowIndex, ColumnIndex)
        666         def __call__(self, RowIndex=defaultNamedOptArg, ColumnIndex=defaultNamedOptArg):
        667                 return self._ApplyTypes_(0, 2, (12, 0), ((12, 17), (12, 17)), '__call__', None,RowIndex
    --> 668                         , ColumnIndex)
        669
        670         def __str__(self, *args):

    D:\Miniconda3\envs\jupyter\lib\site-packages\win32com\client\__init__.py in _ApplyTypes_(self, dispid, wFlags, retType, argTypes, user, resultCLSID, *args)
        457         def _ApplyTypes_(self, dispid, wFlags, retType, argTypes, user, resultCLSID, *args):
        458                 return self._get_good_object_(
    --> 459                         self._oleobj_.InvokeTypes(dispid, 0, wFlags, retType, argTypes, *args),
        460                         user, resultCLSID)
        461

    com_error: (-2147352567, 'Exception occurred.', (0, None, None, None, 0, -2146827284), None)
    --------------------------------------------------------------
    com_error                                 Traceback (most recent call last)
    <ipython-input-71-1ff4d126bb3c> in <module>()
    ----> 1 Range('E1').value = Range('A1').vertical.value

    D:\Miniconda3\envs\jupyter\lib\site-packages\xlwings\main.py in value(self, data)
        847     @value.setter
        848     def value(self, data):
    --> 849         conversion.write(data, self, self._options)
        850
        851     @property

    D:\Miniconda3\envs\jupyter\lib\site-packages\xlwings\conversion\__init__.py in write(value, rng, options)
         20     pipeline = accessors.get(convert, convert).router(value, rng, options).writer(options)
         21     ctx = ConversionContext(rng=rng, value=value)
    ---> 22     pipeline(ctx)
         23     return ctx.value

    D:\Miniconda3\envs\jupyter\lib\site-packages\xlwings\conversion\framework.py in __call__(self, *args, **kwargs)
         64     def __call__(self, *args, **kwargs):
         65         for stage in self:
    ---> 66             stage(*args, **kwargs)
         67
         68

    D:\Miniconda3\envs\jupyter\lib\site-packages\xlwings\conversion\standard.py in __call__(self, ctx)
         53                 ctx.range.col1,
         54                 row2,
    ---> 55                 col2
         56             ), ctx.value)
         57

    D:\Miniconda3\envs\jupyter\lib\site-packages\xlwings\_xlwindows.py in get_range_from_indices(xl_sheet, first_row, first_column, last_row, last_column)
        283 def get_range_from_indices(xl_sheet, first_row, first_column, last_row, last_column):
        284     return xl_sheet.Range(xl_sheet.Cells(first_row, first_column),
    --> 285                           xl_sheet.Cells(last_row, last_column))
        286
        287

    C:\Users\MA17151\AppData\Local\Temp\gen_py\3.4\00020813-0000-0000-C000-000000000046x0x1x6\Range.py in __call__(self, RowIndex, ColumnIndex)
        666         def __call__(self, RowIndex=defaultNamedOptArg, ColumnIndex=defaultNamedOptArg):
        667                 return self._ApplyTypes_(0, 2, (12, 0), ((12, 17), (12, 17)), '__call__', None,RowIndex
    --> 668                         , ColumnIndex)
        669
        670         def __str__(self, *args):

    D:\Miniconda3\envs\jupyter\lib\site-packages\win32com\client\__init__.py in _ApplyTypes_(self, dispid, wFlags, retType, argTypes, user, resultCLSID, *args)
        457         def _ApplyTypes_(self, dispid, wFlags, retType, argTypes, user, resultCLSID, *args):
        458                 return self._get_good_object_(
    --> 459                         self._oleobj_.InvokeTypes(dispid, 0, wFlags, retType, argTypes, *args),
        460                         user, resultCLSID)
        461

    com_error: (-2147352567, 'Exception occurred.', (0, None, None, None, 0, -2146827284), None)
    

  
However, if I use the options method, I can successfully copy/paste the one column of data:  
    
    Range('E1').value = Range('A1').vertical.options(ndim=2).value
    
If I copy / paste more than one column, the options-less method works ok.  
For example, this works:
    
    Range('E1').value = Range('A1:C1').vertical.value
    
Only when I copy / paste one column of data without the options method, do I get the error.

Should the options-less method work with one column of data or is using the options method the preferred way?  I prefer not to use the options method since it is easier to remember the options-less method.  Thanks!

**Comment by fzumstein** [comment]
I can't replicate the issue. Can you try to delete (or rename) this directory: `C:\Users\MA17151\AppData\Local\Temp\gen_py\3.4` and see if it works then? Note that currently, 1d arrays still require `ndim=2` to preserve the column orientation, but it should definitely not throw you an error if you leave it away.

**Comment by sdementen** [comment]
The issue could be that the vertical range has more than 16384 items which
is the max columns allowed in excel 2010.

**Comment by pybokeh** [comment]
Deleting or renaming the cache folder did not fix the problem.  It appears it is a size issue.  With a smaller data set, it worked ok, but the orientation was not correct.  I still had to resort to using the options() method to get the correct orientation.

**Comment by fzumstein** [comment]
Yeah, currently 1d-arrays (whether columns or rows) are interpreted by xlwings as simple lists or 1d-numpy-arrays, if not explicitly set to `ndim=2`, see [here](http://docs.xlwings.org/en/stable/datastructures.html#lists).
Great notebook, by the way! However, Excel constants are also directly available in xlwings:

    from xlwings import constants
    constants.HAlign.xlHAlignCenter

**Comment by pybokeh** [comment]
@fzumstein Thanks for the tip about getting Excel constants directly and also for the work done with xlwings!

**Comment by DougAJ4** [comment]
With a bit of experimenting I found the best option seems to be:
`
Range('D1').options(transpose=True).value = Range('A1').vertical.value
`
This works with a full column of values (1,048,576 values), at least on my system.  I'd be interested to know if it works for sdementen as well.

As noted in the earlier thread (https://github.com/ZoomerAnalytics/xlwings/issues/397), performance remains an issue.  Copying 1 million+ rows takes of the order of 10 seconds, compared with less than 1 second using my old ExcelPython based routines (which still work when called from xlwings).

**Comment by sdementen** [comment]
Hi Doug
I have in general no issues when assigning large arrays to a range.value (as long as it does fit in excel in terms of #cols and  #rows)
BTW, could you try to copy a 1mio rows x 5 columns ? I hit a memory error while reading the data on my machine (I have made a very draft PR for that but which needs a clean rework)  
I have also issues when returning arrays of more than 65536 rows via an UDF (but it works if I simply assign to a range.value)

**Comment by DougAJ4** [comment]
Yes, 5 complete columns copied without a problem, other than taking about 30 seconds to complete.

I tried:
`Range('K1').value = Range('A1').vertical.value`
and
`Range('K1').value = Range('A1').table.value`

Both worked much the same.

**Comment by sdementen** [comment]
I can copy 3 columns of 1048576 rows full of random numbers with
`Range("A1:C1").vertical.xl_range.Value`
but on 4 columns with
`Range("A1:D1").vertical.xl_range.Value`
I get the exception.
`pywintypes.com_error: (-2147024882, 'Not enough storage is available to complete this operation.', None, None)`

Running Excel 2010 on Win7Pro

**Comment by DougAJ4** [comment]
That syntax worked for me for up to at least 5 columns.

But I re-checked the two I ran previously, and only .table worked.  When I ran:
` Range('K1').value = Range('A1').vertical.value `
I got a VBA error at:
Py.Exec "" & PythonCommand & ""

That was with optimized_connection set to True.
With optimized_connection set to False both .vertical and .table failed (with a long list of Python errors), but  Range("A1:D1").vertical.xl_range.Value still works.

I'm running Excel 2016 and Windows 10, with 8GB RAM.  The OS is 64 bit, but Office is 32 bit.
